community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/IndexingService.java

```diff
@@ -335,18 +339,162 @@ private String indexUserDescription( final IndexDescriptor descriptor,
         return String.format( "%s [provider: %s]", userDescription, providerDescriptor.toString() );
     }
 
-    public void updateIndexes( IndexUpdates updates )
+    public ValidatedIndexUpdates validate( IndexUpdates updates )
     {
+        IndexUpdateMode updateMode = null;
         if ( state == State.RUNNING )
         {
-            try ( IndexUpdaterMap updaterMap = indexMapReference.getIndexUpdaterMap( IndexUpdateMode.ONLINE ) )
+            updateMode = IndexUpdateMode.ONLINE;
+        }
+        else if ( state == State.STARTING )
+        {
+            updateMode = IndexUpdateMode.RECOVERY;
+        }
+
+        if ( updateMode != null )
+        {
+            IndexUpdaterMap updaterMap = indexMapReference.getIndexUpdaterMap( updateMode );
+            boolean updaterMapShouldBeClosed = true;
+            try
             {
-                applyUpdates( updates, updaterMap );
+                Map<IndexDescriptor,List<NodePropertyUpdate>> updatesByIndex = groupUpdatesByIndexDescriptor(
+                        updates, updaterMap );
+
+                if ( updatesByIndex.isEmpty() )
+                {
+                    return newValidatedIndexUpdates( updates );
```

Comment:
Recovery procedure also uses this path and it needs `ValidatedIndexUpdates#changedNodeIds` method to return correct set regardless of `updatesByIndex` being empty or not. I thought about splitting recovery & commit paths here so that `NeoStoreTransaction` would decide which one to follow based on it's state, but maybe it is worth doing only in 2.2 to not complicate changes in main branch?